core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java

```diff
@@ -129,8 +152,83 @@ public float getProgress() throws IOException, InterruptedException {
     }
 
     @Override
+    public boolean nextKeyValue() throws IOException, InterruptedException {
+      return delegate.nextKeyValue();
+    }
+
+    @Override
     public void close() throws IOException {
       delegate.close();
     }
   }
+
+  private static class TrackingTextRecordReader extends TrackingRecordReader<LongWritable, Text> {
+
+    private TrackingTextRecordReader(RecordReader<LongWritable, Text> delegate, @Nullable String pathField,
+                                     String path) {
+      super(delegate, pathField, path, getTextOutputSchema(pathField));
+    }
+
+    public StructuredRecord getCurrentValue(StructuredRecord.Builder recordBuilder) throws IOException,
+      InterruptedException {
+      LongWritable key = delegate.getCurrentKey();
+      Text text = delegate.getCurrentValue();
+
+      recordBuilder.set("offset", key.get());
+      recordBuilder.set("body", text.toString());
+      return recordBuilder.build();
+    }
+  }
+
+  private static class TrackingAvroRecordReader extends TrackingRecordReader<AvroKey<GenericRecord>, NullWritable> {
+    private static AvroToStructuredTransformer recordTransformer;
+    private Job job;
+
+    private TrackingAvroRecordReader(RecordReader<AvroKey<GenericRecord>, NullWritable> delegate,
+                                     @Nullable String pathField, String path, String schema, Job job)
+      throws IOException {
+      super(delegate, pathField, path, Schema.parseJson(schema));
+      this.job = job;
+    }
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException,
+      InterruptedException {
+      super.initialize(split, context);
+      recordTransformer = new AvroToStructuredTransformer();
+    }
+
+    public StructuredRecord getCurrentValue(StructuredRecord.Builder recordBuilder) throws IOException,
+      InterruptedException {
+      AvroJob.setOutputKeySchema(job, new org.apache.avro.Schema.Parser().parse(schema.toString()));
+      StructuredRecord ret = recordTransformer.transform(recordBuilder, delegate.getCurrentKey().datum(), schema,
+                                                         pathField);
+      return ret;
+    }
+  }
+
+  private static class TrackingParquetRecordReader extends TrackingRecordReader<Void, GenericRecord> {
+    private AvroToStructuredTransformer recordTransformer;
+    private Job job;
+
+    private TrackingParquetRecordReader(RecordReader<Void, GenericRecord> delegate, @Nullable String pathField,
+                                        String path, String schema, Job job) throws IOException {
+      super(delegate, pathField, path, Schema.parseJson(schema));
```

Comment:
parsing should not occur in the constructor, where it can throw an IOException. It should be parsed by the caller, with only a valid schema object passed into each different type of record reader.